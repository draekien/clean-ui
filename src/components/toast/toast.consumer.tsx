import { ToastContext } from './toast.context';

export const ToastConsumer = ToastContext.Consumer;
