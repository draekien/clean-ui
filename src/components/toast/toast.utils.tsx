import { StatusVariant } from '../../types/variants';
import { colors } from '../theme/theme.colors';

export const iconMap: { [key in StatusVariant]: string } = {
  info: 'info',
  success: 'check_circle',
  warning: 'warning',
  error: 'error',
};

export const iconColorMap: { [key in StatusVariant]: keyof typeof colors } = {
  info: 'info-400',
  success: 'success-400',
  warning: 'warning-400',
  error: 'error-400',
};
