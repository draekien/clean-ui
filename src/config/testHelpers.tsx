import React, { ReactNode } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: ReactNode) =>
  render(<ThemeProvider>{element}</ThemeProvider>);
