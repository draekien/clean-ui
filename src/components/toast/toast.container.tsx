/** @jsxImportSource theme-ui */
import React from 'react';
import { Toast } from './toast';
import { ToastConsumer } from './toast.consumer';
import { toastContainerCss } from './toast.styles';

export type ToastPosition = 'top' | 'bottom' | 'right';
export interface ToastContainerProps {
  position?: ToastPosition;
  offset?: string | number | string[] | number[];
}

export const ToastContainer: React.FC<ToastContainerProps> = ({
  position = 'right',
  offset,
  ...rest
}) => (
  <ToastConsumer>
    {(context) => (
      <div
        id="clean_ui_toast_container"
        sx={toastContainerCss({ position, offset })}
        {...rest}>
        {context.toasts.map((toast) => (
          <Toast key={toast.key} openFrom={position} {...toast} />
        ))}
      </div>
    )}
  </ToastConsumer>
);
