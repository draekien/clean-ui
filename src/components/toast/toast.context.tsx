import { createContext, useContext } from 'react';
import { ToastProps } from './toast';

export interface ToastContextProps {
  toasts: ToastProps[];
  open: Function;
  close: Function;
  closeAll: Function;
  success: Function;
  error: Function;
  warning: Function;
  info: Function;
}

export const ToastContext = createContext<ToastContextProps>({
  toasts: [],
  open: () => {},
  close: () => {},
  closeAll: () => {},
  success: () => {},
  error: () => {},
  warning: () => {},
  info: () => {},
});

export const useToastContext = () => useContext(ToastContext);
