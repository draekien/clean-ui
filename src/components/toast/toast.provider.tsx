import React, { useState } from 'react';
import { ToastProps } from './toast';
import { ToastContext, ToastContextProps } from './toast.context';

export interface ToastProviderProps {
  initialState?: ToastProps[];
  defaultProps?: ToastProps;
}

export const ToastProvider: React.FC<ToastProviderProps> = ({
  initialState = [],
  defaultProps = [],
  children,
}) => {
  const [toasts, setToasts] = useState(initialState);

  let count = 0;
  const getKey = () => `${Date.now()}_${count++}`;

  const open = (props: ToastProps) => {
    const active = [...toasts];
    const newToast = Object.assign({}, defaultProps, props);

    if (!newToast.key) newToast.key = getKey();
    active.push(newToast);

    setToasts(active);
  };

  const close = (key: string) => {
    const active = [...toasts];

    active.forEach((toast, index) => {
      if (toast.key === key) active[index].open = false;
    });

    setToasts(active);
  };

  const closeAll = () => {
    const active = [...toasts];

    active.forEach((toast) => {
      toast.open = false;
    });

    setToasts(active);
  };

  const success = (props: ToastProps) => {
    open(Object.assign({}, props, { variant: 'success' }));
  };

  const warning = (props: ToastProps) => {
    open(Object.assign({}, props, { variant: 'warning' }));
  };

  const error = (props: ToastProps) => {
    open(Object.assign({}, props, { variant: 'error' }));
  };

  const info = (props: ToastProps) => {
    open(Object.assign({}, props, { variant: 'info' }));
  };

  const context: ToastContextProps = {
    toasts,
    open,
    close,
    closeAll,
    success,
    warning,
    error,
    info,
  };

  return <ToastContext.Provider value={context}>{children}</ToastContext.Provider>;
};
