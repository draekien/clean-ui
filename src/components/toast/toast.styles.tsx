import { ThemeUIStyleObject } from '@theme-ui/css';
import { keyframes } from '@emotion/react';
import { StatusVariant } from '../../types/variants';
import { ToastContainerProps, ToastPosition } from './toast.container';

const slideLeft = keyframes`
0% {
  transform: translateX(100%);
}
100% {
  transform: translateX(0%);
}`;

const slideUp = keyframes`
0% {
  transform: translateY(200%);
}
100% {
  transform: translateY(0%);
}`;

const slideDown = keyframes`
0% {
  transform: translateY(-200%);
}
100% {
  transform: translateY(0%);
}
`;

export interface ToastCssProps {
  variant: StatusVariant;
  clickable?: boolean;
  position?: ToastPosition;
}

export const toastContainerCss = ({
  position,
  offset,
}: ToastContainerProps): ThemeUIStyleObject => {
  const css: ThemeUIStyleObject = {
    position: 'fixed',
    display: 'flex',
    flexFlow: 'column',
    zIndex: 'toast',
    pointerEvents: 'none',
    right: 'calc(50% - 11rem)',
  };

  switch (position) {
    case 'bottom':
      css.bottom = offset || 0;
      css.top = 0;
      css.justifyContent = 'flex-end';
      break;
    case 'top':
      css.top = offset || 0;
      css.bottom = 0;
      break;
    default:
      css.right = 0;
      css.top = offset || 0;
      css.bottom = 0;
      break;
  }

  return css;
};

export const toastWrapperCss = ({
  variant,
  clickable,
  position,
}: ToastCssProps): ThemeUIStyleObject => {
  const css: ThemeUIStyleObject = {
    backgroundColor: `${variant}-000`,
    borderLeft: '4px solid',
    borderLeftColor: `${variant}-400`,
    borderRadius: 'md',
    boxShadow: 'sm',
    margin: '1rem',
    opacity: 0,
    pointerEvents: 'all',
    position: 'relative',
    transition: 'all 300ms',
    width: '20rem',
    zIndex: 'toast',
  };

  if (clickable) {
    css.cursor = 'pointer';
  }

  switch (position) {
    case 'top':
      css.top = 0;
      css.animation = `${slideDown} 700ms`;
      break;
    case 'bottom':
      css.bottom = 0;
      css.animation = `${slideUp} 700ms`;
      break;
    default:
      css.right = 0;
      css.animation = `${slideLeft} 700ms`;
  }

  return css;
};

export const toastWrapperTransitions: any = {
  entering: { opacity: 0 },
  entered: { opacity: 1 },
  exiting: {
    opacity: 0,
    height: '0 !important',
    overflow: 'hidden',
  },
  exited: {
    opacity: 0,
    height: '0 !important',
    overflow: 'hidden',
  },
};

export const toastVariantIconCss: ThemeUIStyleObject = {
  margin: 'md',
  position: 'absolute',
};

export const toastTitleCss = ({ variant }: ToastCssProps): ThemeUIStyleObject => ({
  color: `${variant}-400`,
  minHeight: '2rem',
  pt: 'md',
  pr: 'xxl',
  pb: 0,
  pl: '3.5rem',
  variant: 'text.subtitle',
});

export const toastMessageCss = ({ variant }: ToastCssProps): ThemeUIStyleObject => ({
  minHeight: '2rem',
  p: '0 2.5rem 1rem 3.5rem',
  pt: 'xs',
  variant: 'text.small',
  color: `${variant}-300`,
});

export const toastCloseIconCss: ThemeUIStyleObject = {
  position: 'absolute',
  right: 'sm',
  top: 'sm',
  cursor: 'pointer',
  color: 'b-600',
};
