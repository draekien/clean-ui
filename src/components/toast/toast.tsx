// eslint-disable-next-line tsdoc/syntax
/** @jsxImportSource theme-ui */
import React, { useState, useEffect, SyntheticEvent, MouseEventHandler } from 'react';
import { ToastPosition } from './toast.container';
import { StatusVariant } from '../../types/variants';
import { Transition } from 'react-transition-group';
import { Icon } from '../icon/icon';
import { iconColorMap, iconMap } from './toast.utils';
import {
  toastCloseIconCss,
  toastMessageCss,
  toastTitleCss,
  toastVariantIconCss,
  toastWrapperCss,
  toastWrapperTransitions,
} from './toast.styles';

export interface ToastProps extends React.HTMLAttributes<HTMLDivElement> {
  key?: string;
  title?: string;
  message?: string;
  openFrom?: ToastPosition;
  variant?: StatusVariant;
  duration?: number;
  open?: boolean;
  onClose?: (e: SyntheticEvent<HTMLDivElement, Event> | undefined) => any;
  onClick?: (e: SyntheticEvent) => any;
}

export const Toast: React.FC<ToastProps> = ({
  title,
  message,
  openFrom = 'right',
  variant = 'success',
  duration = 6.5,
  open = true,
  onClose,
  onClick,
  ...rest
}) => {
  const [openState, setOpenState] = useState(open);
  const [openTime, setOpenTime] = useState(Date.now());

  let timer: number;
  let toastElement: HTMLDivElement | null;

  useEffect(() => {
    if (toastElement) toastElement.style.height = toastElement.scrollHeight + 'px';
  }, [message]);

  useEffect(() => {
    setOpenState(open);
  }, [open]);

  const handleClose: MouseEventHandler<HTMLDivElement> = (e) => {
    clearTimeout(timer);
    setOpenState(false);
    onClose && onClose(e);
  };

  const handleClick: MouseEventHandler<HTMLDivElement> = (e) => {
    onClick && onClick(e);
  };

  const setTimer = (isMouseLeave = false) => {
    if (isMouseLeave) setOpenTime(Date.now());

    if (duration) {
      const durationMs = duration * 1000;
      const elapsed = Date.now() - openTime;
      const timeout = durationMs - elapsed;
      timer = setTimeout(handleClose, timeout);
    }
  };

  return (
    <Transition appear in={openState} timeout={300} unmountOnExit>
      {(state) => (
        <div
          ref={(el) => (toastElement = el)}
          sx={{
            ...toastWrapperCss({ variant, clickable: !!onClick, position: openFrom }),
            ...toastWrapperTransitions[state],
          }}
          onClick={handleClick}
          onMouseEnter={() => clearTimeout(timer)}
          onMouseLeave={() => setTimer(true)}
          {...rest}>
          <Icon
            sx={toastVariantIconCss}
            name={iconMap[variant]}
            color={iconColorMap[variant]}
          />
          <div sx={toastTitleCss({ variant })}>{title}</div>
          <div sx={toastMessageCss({ variant })}>{message}</div>
          <Icon sx={toastCloseIconCss} name="close" onClick={handleClose} size="small" />
        </div>
      )}
    </Transition>
  );
};

Toast.displayName = 'Toast';

export * from './toast.consumer';
export * from './toast.provider';
export * from './toast.context';
export * from './toast.container';
