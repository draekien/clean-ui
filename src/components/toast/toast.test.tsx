import React from 'react';
import { themedRender } from '../../helpers/testHelpers';
import { Toast } from './toast';

describe('Toast component', () => {
  it('Should match snapshot', () => {
    const { container } = themedRender(
      <Toast title="toast" message="message" onClose={() => {}} onClick={() => {}} />
    );

    expect(container).toMatchSnapshot();
  });
});
