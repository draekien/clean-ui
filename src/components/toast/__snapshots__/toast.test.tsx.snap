// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast component Should match snapshot 1`] = `
<div>
  <div
    class="css-21jisu"
    style="height: 0px;"
  >
    <i
      class="material-icons css-35gkst-Icon"
    >
      check_circle
    </i>
    <div
      class="css-txoam9"
    >
      toast
    </div>
    <div
      class="css-664lrf"
    >
      message
    </div>
    <i
      class="material-icons css-1op4n3b-Icon"
    >
      close
    </i>
  </div>
</div>
`;
